src/atomic.rs

```diff
@@ -396,64 +288,63 @@ impl<T> Atomic<T> {
     /// # Examples
     ///
     /// ```
-    /// use crossbeam_epoch::{self as epoch, Atomic, Owned};
+    /// use crossbeam_epoch::{self as epoch, Atomic, Owned, Shared};
     /// use std::sync::atomic::Ordering::SeqCst;
     ///
     /// let a = Atomic::new(1234);
     ///
     /// let guard = &epoch::pin();
     /// let mut curr = a.load(SeqCst, guard);
-    /// let res = a.compare_and_set_owned(curr, Owned::new(5678), SeqCst, guard);
+    /// let res1 = a.compare_and_set(curr, Shared::null(), SeqCst, guard);
+    /// let res2 = a.compare_and_set(curr, Owned::new(5678), SeqCst, guard);
     /// ```
-    pub fn compare_and_set_owned<'g, O>(
+    pub fn compare_and_set<'g, O, P>(
         &self,
-        current: Ptr<T>,
-        new: Owned<T>,
+        current: Shared<T>,
+        new: P,
         ord: O,
         _: &'g Guard,
-    ) -> Result<Ptr<'g, T>, (Ptr<'g, T>, Owned<T>)>
+    ) -> Result<Shared<'g, T>, (Shared<'g, T>, P)>
     where
         O: CompareAndSetOrdering,
+        P: Pointer<T>,
```

Comment:
This looks kinda messy at the moment, but with in-band lifetimes (see [here](https://github.com/rust-lang/rust/pull/46051)) and impl Trait in argument position (see [here](https://github.com/rust-lang/rust/pull/45918)), this signature will be simplified to:

```rust
pub fn compare_and_set(
    &self,
    current: Shared<T>,
    new: impl Pointer<T>,
    ord: impl CompareAndSetOrdering,
    _: &'g Guard,
) -> Result<Shared<'g, T>, (Shared<'g, T>, P)>
```

Reply:
The return type stands out as really sigil-heavy and could potentially be changed to `Result<Shared<'g, T>, CasError<'g, T, P>>` with fields like `curr` and `new` in `CasError` or even entirely to `CasResult<'g, T, P>`. That said, the current interface uses standard types and is understandable without looking up more custom types, so i'm not so sure. The linked RFCs will definitely improve things.

Reply:
I'm in favor of `CasError<'g, T, P>` because it's easy to mistake the tuple `(previous, new)` for `(new, previous)`.